File: clojure/src/project_euler/sol/p0056.clj
(ns project-euler.sol.p0056
  (:require
   [project-euler.lib.math :as my-math]
   [project-euler.lib.util :as util]))

(defn- aux
  [n upper]
  (loop [ans 0
         prod (my-math/pow n upper)
         cnt upper]
    (let [prod-digits (util/digits prod)]
      (if (or (zero? cnt)
              (< (* 9 (count prod-digits)) ans))
        ans
        (recur (max (apply + prod-digits) ans)
               (quot prod n)
               (dec cnt))))))

;;; assume that x = 10 * n
;;;   x^y = (10 * n)^y = 10^y * n^y, so sum_of_digits(x^y) = sum_of_digits(n^y)
;;;   we can skip to check multiples of ten in this problem.
(defn solve
  ([]
   (solve 100))
  ([upper]
   {:pre [(> upper 1)]}
   (let [xf (comp (remove #(zero? (mod % 10))) ; Skip when multiple of ten.
                  (map #(aux % (dec upper))))]
     (apply max (into [] xf (range (dec upper) 0 -1))))))

File: julia/Euler/src/bin/p0056.jl

# project euler: problem 56

module Prob0056

function solve_0056()
    answer = 0
    for a = reverse(1:BigInt(99))
        # assume that x = 10 * n
        # x^y = (10 * n)^y = 10^y * n^y, so sum(digits(x^y)) = sum(digits(n^y))
        # we can skip to check multiples of ten in this problem.
        if a % 10 == 0
            continue
        end
        for b = reverse(1:BigInt(99))
            p = a ^ b
            if length(string(p)) * 9 < answer
                break
            end
            answer = max(sum(digits(p)), answer)
        end
    end
    answer
end

end #module

using .Prob0056: solve_0056
export solve_0056
